torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,394 @@
+            for param in group["params"]:
+                if param.grad is not None:
+                    if (not allow_fp16) and param.grad.dtype == torch.float16:
+                        raise ValueError("Attempting to unscale FP16 gradients.  If you want to check for "
+                                         "infs/nans without unscaling, use optimizer.check_infs() instead.")
+                    else:
+                        torch._amp_non_finite_check_and_unscale_(param.grad,
+                                                                 per_device_found_inf.get(param.grad.device),
+                                                                 per_device_inv_scale.get(param.grad.device))
+
+        return per_device_found_inf._per_device_tensors
+
+    def unscale(self, optimizer):
+        """
+        Divides ('unscales') the optimizer's gradient tensors by the scale factor.
+
+        If :meth:`unscale` is not called explicitly then gradients will be
+        automatically unscaled during :meth:`step`.
+
+        :meth:`unscale` can be called explicitly to manipulate a network's
+        gradients after backward and before :meth:`step`.
+
+        Simple example, using :meth:`unscale` to enable clipping of unscaled gradients::
+
+            ...
+            scaler.scale(loss).backward()
+            scaler.unscale(optimizer)
+            torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)
+            scaler.step(optimizer)
+            scaler.update()
+
+        If this instance of :class:`AmpScaler` is not enabled, :meth:`unscale` is a no-op.
+
+        Arguments:
+            optimizer (torch.optim.Optimizer):  Optimizer that owns the gradients to be unscaled.
+
+        .. note::
+            :meth:`unscale` does not incur a CPU-GPU sync.
+
+        .. warning::
+            :meth:`unscale` should only be called once per optimizer per step,
+            and only after all gradients for that optimizer's owned parameters
+            have been accumulated.
+        """
+        if not self._enabled:
+            return
+
+        assert self._scale is not None, self._scale_not_initialized_error("unscale")
+
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+
+        if "unscaled" in optimizer_state:
+            raise RuntimeError("unscale() has already been called on this optimizer this iteration.")
+
+        # FP32 division can be imprecise for certain compile options, so we carry out the reciprocal in FP64.
+        inv_scale = self._scale.double().reciprocal().float()
+        found_inf = torch.full((1,), 0.0, dtype=torch.float32, device=self._scale.device)
+
+        optimizer_state["found_inf_per_device"] = self._unscale_grads(optimizer, inv_scale, found_inf)
+        optimizer_state["unscaled"] = True
+
+    def step(self, optimizer, *args, **kwargs):
+        """
+        Carry out a scaling-safe ``optimizer.step()``.  "Scaling-safe" means two things:
```

Comment:
Mike: it's great that you explain this but "scaling-safe" is still not a great name.